src/pdmat.jl

```diff
@@ -1,23 +1,54 @@
 """
 Full positive definite matrix together with a Cholesky factorization object.
 """
-struct PDMat{T<:Real,S<:AbstractMatrix} <: AbstractPDMat{T}
+struct PDMat{T<:Real,S<:AbstractMatrix{T}} <: AbstractPDMat{T}
     mat::S
     chol::Cholesky{T,S}
 
-    PDMat{T,S}(m::AbstractMatrix{T},c::Cholesky{T,S}) where {T,S} = new{T,S}(m,c)
-end
-
-function PDMat(mat::AbstractMatrix,chol::Cholesky{T,S}) where {T,S}
-    d = LinearAlgebra.checksquare(mat)
-    if size(chol, 1) != d
-        throw(DimensionMismatch("Dimensions of mat and chol are inconsistent."))
+    function PDMat{T,S}(m::AbstractMatrix, c::Cholesky) where {T,S}
+        d = LinearAlgebra.checksquare(m)
+        if size(c, 1) != d
+            throw(DimensionMismatch("Dimensions of mat and chol are inconsistent."))
+        end
+        # in principle we might want to check that `c` is a Cholesky factorization of `m`,
+        # but that's slow
+        return new{T,S}(m,c)
     end
-    PDMat{T,S}(convert(S, mat), chol)
 end
+function PDMat{T}(m::AbstractMatrix, c::Cholesky) where T
+    c = convert(Cholesky{T}, c)
+    return PDMat{T,mattype(c)}(m, c)
+end
+PDMat(mat::AbstractMatrix,chol::Cholesky{T,S}) where {T,S} = PDMat{T,S}(mat, chol)
+
+# Construction from another PDMat
+PDMat{T,S}(pdm::PDMat{T,S}) where {T,S} = pdm
```

Comment:
I wonder if the constructor should return an actual copy given https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#Mutable-collections. On the other hand, based on the PR you linked to LinearAlgebra doesn't seem to follow this convention?

Reply:
It's technically not a mutable collection: `pdm.mat = rand(3,3)` will fail.